src/dune_rules/pkg_rules.ml

```diff
@@ -1359,32 +1359,52 @@ module DB = struct
     entry.pkg_digest
   ;;
 
-  let of_ctx ctx ~allow_sharing =
```

Comment:
Since you have 2 parameters rather than one, I think using `Memo.create` is cleaner than using `Per_context`. Splitting the tables like this does not really buy you much.